### PR TITLE
Don't use server commands for scheduling plugin unloads.

### PR DESCRIPTION
**Pull request by dvander**
If a plugin is currently active on the call stack, we can't unload immediately unload it since we would crash. The old code to deal with this used server commands to unload the plugin. That's pretty sketchy for a bunch of reasons.

The new method below adds the ability to attack one-off callbacks to IServerGameDLL::Think(), and then we use that to flush plugin unloads instead.

### CODE REVIEW

File: core/sourcemod.cpp [hunk 1]
```diff
@@ -540,6 +543,7 @@ void SourceModBase::ShutdownServices()
 
 	SH_REMOVE_HOOK(IServerGameDLL, LevelShutdown, gamedll, SH_MEMBER(this, &SourceModBase::LevelShutdown), false);
 	SH_REMOVE_HOOK(IServerGameDLL, GameFrame, gamedll, SH_MEMBER(&g_Timers, &TimerSystem::GameFrame), false);
+	SH_REMOVE_HOOK(IServerGameDLL, Think, gamedll, SH_STATIC(*logicore.OnThink), false);
```
Comment (VoiDeD) [inline core/sourcemod.cpp]: Unintentional dereference?

Comment (dvander) [inline core/sourcemod.cpp]: Yup, thanks - though it shouldn't matter, it'll decay back into a pointer if it needs to.

**Comment by asherkin** [comment]
tumblr_nsxrrobang1r0q799o1_500

**Comment by dvander** [comment]
This image is terrifying.

**Comment by dvander** [comment]
Pushed via command-line.

**Pull request closed without merging**